src/login/logind-session.c

```diff
@@ -170,23 +170,16 @@ void session_set_user(Session *s, User *u) {
         user_update_last_session_timer(u);
 }
 
-int session_set_leader(Session *s, pid_t pid) {
-        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+int session_set_leader(Session *s, PidRef leader) {
         int r;
 
         assert(s);
+        assert(pidref_is_set(&leader));
 
-        if (!pid_is_valid(pid))
-                return -EINVAL;
-
-        if (s->leader.pid == pid)
+        if (s->leader.pid == leader.pid)
                 return 0;
```

Comment:
memory mgmt is borked here now. you donated the PidRef to session_set_leader() in the earlier call, but here you don't release it if you exit early.

you basically have three options here:

1. donate the pidref to this function in both error and success. in that case i'd immediately assign the parameter to a local variable that uses `_cleanup_` to release the thing if the scope is left. i.e. rename the param `_pidref` and then do `_cleanup_(pidref_done) PidRef pidref = _pidref;` or so.

2. donate it only on success, in which case you make sure that if you exit >= 0 you use TAKE_PIDREF() once, but on the failure paths you just leave the thing as it is.

3. you make a clone of the passed fd, i.e. no dontation takes place. in that case you use pidref_set_pidfd() and pass the pidref.fd of the parameter to it.

Usually when we go for option 1 we indicate that in the function name, i.e. rename it to xyz_consume() or so, since our typical rule is to take possession of things only on success, not on failure.